src/heitech.zer0mqXt.core/Patterns/Socket.cs

```diff
@@ -118,20 +123,27 @@ public Socket(SocketConfiguration configuration)
                     catch (System.Exception ex)
                     {
                         // failure to parse or any other exception leads to a non successful response, which then in turn can be handled on the request side
-                        // todo exception propagation?
                         _configuration.Logger.Log(new ErrorLogMsg($"Responding to [Request:{typeof(T)}] with [Response:{typeof(TResult)}] did fail: " + ex.Message));
                         response = new RequestReplyMessage<TResult>(_configuration, default(TResult), success: false);
                     }
                     // try send response with timeout
                     bool noTimeout = rsSocket.TrySendMultipartMessage(_configuration.TimeOut, response);
                     if (!noTimeout)
                         _configuration.Logger.Log(new ErrorLogMsg($"Responding to [Request:{typeof(T)}] with [Response:{typeof(TResult)}] timed-out after {_configuration.TimeOut}"));
+
+                    if (respondOnce)
+                        runReponder = false;
                 }
-            });
+            }, UseBestToken(cancellationToken));
```

Comment:
I'm not sure: did you mean to enable cancellation only before the function has actually started - or is there a `cancellationToken.ThrowIfCancellationRequested();` missing, perhaps at the beginning of the `while` loop - or alternatively graceful cancellation via additionally checking for `while (runReponder && !cancellationToken.IsCancellationRequested)`?

Reply:
Maybe I am using this incorrectly but the intention is to kill the server with this.
In other words, if the Client creating this server adds a CancellationToken to this method, it can control the Task Execution if I am not mistaken.

[see here in the microsoft docs for reference](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/how-to-cancel-a-task-and-its-children)

Reply:
I haven't used `Task.Run` in a while, but IIRC any user-defined Operation, once startet, is _never_ "automatically" canceled - it's completion may no longer be observed with e.g. a `Task.WhenAny`, but even then the operation continues to run.

And I believe it's the same thing with `Task.Run`: if the custom operation has not yet started, and the `CancellationToken` passed to `Task.Run(Action, CancellationToken)` was canceled, then it would indeed cancel the resulting `Task`. But after the `Task` has started to run, we need to "cooperatively" cancel it, either gracefully by letting the `Action` complete via `cancellationToken.IsCancellationRequested` in the loop's condition, or by throwing a `OperationCanceledException` via `cancellationToken.ThrowIfCancellationRequested()` within the custom operation.